core/src/identifier.rs

```diff
@@ -1,35 +1,107 @@
 //! Define the type of an identifier.
 use once_cell::sync::Lazy;
 use serde::{Deserialize, Serialize};
-use std::fmt::{self, Debug};
-use std::hash::Hash;
+use std::{
+    fmt::{self, Debug},
+    hash::Hash,
+};
 
 use crate::{position::TermPos, term::string::NickelString};
 
 simple_counter::generate_counter!(GeneratedCounter, usize);
 static INTERNER: Lazy<interner::Interner> = Lazy::new(interner::Interner::new);
 
-#[derive(Clone, Copy, Deserialize, Serialize)]
+/// An interned string.
+//
+// Implementation-wise, this is just a wrapper around interner::Symbol that uses a hard-coded,
+// static `Interner`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
 #[serde(into = "String", from = "String")]
-pub struct Ident {
-    symbol: interner::Symbol,
-    pub pos: TermPos,
-    generated: bool,
+pub struct Symbol(interner::Symbol);
+
+impl Symbol {
+    pub fn new(s: impl AsRef<str>) -> Self {
+        Self(INTERNER.intern(s.as_ref()))
+    }
+
+    /// Return the string representation of this symbol.
+    pub fn label(&self) -> &str {
+        INTERNER.lookup(self.0)
+    }
+
+    pub fn into_label(self) -> String {
+        self.label().to_owned()
+    }
+}
+
+impl fmt::Display for Symbol {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.label())
+    }
+}
+
+impl From<Symbol> for Ident {
+    fn from(symbol: Symbol) -> Self {
+        Ident {
+            symbol,
+            pos: TermPos::None,
+            generated: symbol.label().starts_with(GEN_PREFIX),
+        }
+    }
 }
 
-impl Debug for Ident {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        f.debug_struct("Ident")
-            .field("label", &self.label())
-            .finish()
+impl PartialOrd for Symbol {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.label().partial_cmp(other.label())
+    }
+}
+
+impl Ord for Symbol {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.label().cmp(other.label())
     }
 }
 
+impl From<Symbol> for NickelString {
+    fn from(sym: Symbol) -> Self {
+        sym.to_string().into()
+    }
+}
+
+impl<F> From<F> for Symbol
+where
+    String: From<F>,
+{
+    fn from(val: F) -> Self {
+        Self(INTERNER.intern(String::from(val)))
+    }
+}
+
+#[allow(clippy::from_over_into)]
+impl Into<String> for Symbol {
+    fn into(self) -> String {
+        self.into_label()
+    }
+}
+
+/// An identifier is a [`Symbol`] together with a position.
+///
+/// The position is ignored for equality comparison and hashing; it's mainly
+/// intended for error messages. If you want to combine a symbol with a semantically
+/// meaningful position, use [`RichIdent`] instead.
```

Comment:
Is `RichIdent` planned for a future PR? By the way, this naming scheme is unfortunate (one day, the sooner the better, I will rip `RichTerm` out as well). I think we could use a better convention. Maybe something like `LocIdent` or `IdentLoc`? (so that there's no possible confusion with `XXXPos`, such as `TermPos`, which is "the position of a term", while here it's more like a "located ident"/ident+location).

Reply:
Ah, I had planned it when I wrote the comment, but then I decided that it will live in `nls` instead of `nickel-core` (in a future PR). I'll remove it from the comment.